tests/infra/module_tester.py

```diff
@@ -0,0 +1,141 @@
+# SPDX-FileCopyrightText: (c) 2024 Tenstorrent AI ULC
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from enum import Enum
+from typing import Callable, Sequence, Union
+
+import jax
+
+from .device_runner import DeviceRunner
+from .test_model import TestModel
+from .test_module import TestModule
+from .utils import (
+    compare_allclose,
+    compare_atol,
+    compare_equal,
+    compare_pcc,
+    random_tensor,
+)
+
+
+class TestType(Enum):
+    INFERENCE = "inference"
+    TRAINING = "training"
+
+
+class ComparisonMetric(Enum):
+    EQUAL = "equal"
+    PCC = "pcc"
+    ATOL = "atol"
+    ALLCLOSE = "allclose"
+
+
+class ModuleTester:
+    """
+    Class providing infrastructure for testing test modules.
+
+    Testing consists of comparing output results of running provided test module on
+    different devices. Frequently, run on CPU or GPU is taken as a ground truth
+    ("golden"), while custom device run is compared to it.
+
+    It supports testing inference and training modes.
+    """
+
+    def __init__(
+        self,
+        test_type: TestType = TestType.INFERENCE,
+        comparison_metric: ComparisonMetric = ComparisonMetric.PCC,
+    ) -> None:
+        self._test_type = test_type
+        self._comparison_metric = comparison_metric
+
+    def __call__(self, module: TestModule) -> bool:
+        """
+        The only public method providing testing hook.
+
+        Call tester with passed test module and it will run tests on it.
+        """
+        return self._test(module)
+
+    def _test(self, module: TestModule) -> bool:
+        if self._test_type == TestType.INFERENCE:
+            return self._test_inference(module)
+        else:
+            return self._test_training(module)
```

Comment:
This is a bit different for ops/graphs and for models. For ops/graphs we want to just run the module and verify results, i.e. inference only. For models we want to test both eval and training modes. We should discuss if we want to separate the op tester from model tester completely.